### DISCUSSION TITLE
Proposal: null-propagating `with` syntax

### DISCUSSION THREAD

**Discussion by RenderMichael** [dotnet/csharplang#6575] [Language Ideas]
Pretty straightforward proposal, a natural extension of the existing functionality, and (I assume) fairly low-cost. 

I imagine the most natural syntax looks like 
``` csharp
potentionallyNullRecord with? { NewValue = x }
```
or possibly
``` csharp
potentionallyNullRecord? with { NewValue = x }
```

**Comment by Rekkonnect** [comment]
The syntax choice will be affected if `await?`, `foreach?`, `return?`, `yield return?`, `throw?` and alike get introduced. The `with?` choice provides a null alternative to a keyword that operates on a non-null operand, throwing otherwise. Since I'm personally in favor of getting those alternatives, I'll support the `with?` variant.

**Reply by jnm2** [reply]
Of these, `await?` has so far been given the green light: https://github.com/dotnet/csharplang/issues/35

**Reply by Rekkonnect** [reply]
And if more are to come, the syntax will lean more towards `x with? { ... }` rather than `x? with { ... }`

**Reply by RenderMichael** [reply]
I hope there can be an LDM "null-a-palooza" which goes over all of these. I can't imagine having too much issue with most of these, except maybe `throw?` having limited use (I like to use `Rethrow(this Exception ex) => ExceptionDispatchInfo.Capture(ex).Throw()` instead, which can be null-propagated as a method)

**Reply by jnm2** [reply]
A thing I've picked up is that there would need to be more of a compelling practical motivation than "this would be consistent." Mostly because each change is significant work in terms of interactions with everything around C#: spec work, thinking about how these might impact future C# features, documentation, compiler implementation, maintenance, and bug reports, light bulb suggestions and saved preferences, and so on.

Teams working on C# features are always working on them instead of some other features. A compelling use case is required in order to prioritize. "Can't imagine having too much issue" or other forms of "why not do it" don't end up being compelling in this context.

**Reply by jnm2** [reply]
> I like to use `Rethrow(this Exception ex) => ExceptionDispatchInfo.Capture(ex).Throw()` instead, which can be null-propagated as a method

This is interesting. Do you have use cases where storing the maybe-null exception in an `Exception` variable rather than an `ExceptionDispatchInfo` variable is the right thing to do?

**Reply by RenderMichael** [reply]
> Do you have use cases where storing the maybe-null exception in an `Exception` variable rather than an `ExceptionDispatchInfo` variable is the right thing to do?

My use case for this is an automation framework I've build around NUnit (I'd love to link it but it's an internal company product).
Generally it goes like

``` csharp
private Exception? Exception;
private bool Passed => exception is not null;
private ILogger _logger;

public void AssertCondition()
{
    try
    {
        Assert.That(...);
    }
    catch (AssertionException ex)
    {
        Exception = ex;
    }
    finally
    {
        _logger.LogInformation($"Additional logging out of NUnit's scope. {(Passed ? '✓' : '❌')}")
        Exception?.Rethrow();
    }
}
```

**Reply by jnm2** [reply]
Have you considered exception filters to avoid stack unwinding and rethrowing?

```cs
private Exception? exception;
private bool Passed => exception is not null;
private ILogger _logger;

private bool RecordExceptionWithoutCatching(Exception ex)
{
    exception = ex;
    return false;
}

public void AssertCondition()
{
    try
    {
        Assert.That(...);
    }
    catch (AssertionException ex) when (RecordExceptionWithoutCatching(ex))
    {
    }
    finally
    {
        _logger.LogInformation($"Additional logging out of NUnit's scope. {(Passed ? '✓' : '❌')}");
    }
}
```

**Reply by jnm2** [reply]
Also, be careful about expecting an exception when an NUnit assertion fails. NUnit doesn't throw exceptions for assertion failures inside Assert.Multiple blocks. A more robust way to check for failure is via `TestContext.CurrentContext.Result.Outcome`:

```cs
private bool Passed => TestContext.CurrentContext.Result.Outcome.Status != TestStatus.Failed;
private ILogger _logger;

public void AssertCondition()
{
    try
    {
        Assert.That(...);
    }
    finally
    {
        _logger.LogInformation($"Additional logging out of NUnit's scope. {(Passed ? '✓' : '❌')}");
    }
}
```

**Reply by RenderMichael** [reply]
I considered that, but then the following code
``` csharp
Assert.Multiple(() =>
{
    AssertSomethingFalse();
    AssertSomethingTrue();
}
```

Would end up producing logs like 
```
Additional logging: ❌
Additional logging: ❌
```

when the correct solution would be
```
Additional logging: ❌
Additional logging: ✓
```

If NUnit's CurrentContext had a property like `ResultOfLastAssertionInMultipleBlock` then it would be feasible, but at the moment I just tell internal consumers of the library not to use `Assert.Multiple`

**Reply by jnm2** [reply]
Gotcha. I'd still recommend using a catch filter rather than rethrowing, and then I'm back to not knowing of motivating cases where the maybe-null variable should be `Exception` and not `ExceptionDispatchInfo`.

**Comment by jnm2** [comment]
I could really use `with?`. This pattern is coming up a lot:
```cs
return state with
{
    SomeDialog = state.SomeDialog is null ? null : state.SomeDialog with
    {
        SomeTextBox = state.SomeDialog.SomeTextBox with { Value = newValue },
    },
};
```

I'd like to be able to write this instead:
```cs
return state with
{
    SomeDialog = state.SomeDialog with?
    {
        SomeTextBox = state.SomeDialog.SomeTextBox with { Value = newValue },
    },
};
```

I also want to be able to collapse this structure like <https://github.com/dotnet/csharplang/issues/4587> specifies, but with `?.` also being available:
```cs
return state with { SomeDialog?.SomeTextBox.Value = newValue };
```

**Reply by RenderMichael** [reply]
The latter part goes hand-in-hand with https://github.com/dotnet/csharplang/issues/6045

**Comment by jnm2** [comment]
`with` on nullable objects had a mention from an LDM member at https://github.com/dotnet/csharplang/issues/6045#issuecomment-1292440765.
